client/components/DataManagement/hooks.js

```diff
@@ -0,0 +1,232 @@
+import { useMemo, useState } from 'react';
+import {
+    DATA_MANAGEMENT_TABLE_FILTER_OPTIONS,
+    DATA_MANAGEMENT_TABLE_SORT_OPTIONS,
+    TABLE_SORT_ORDERS,
+    TEST_PLAN_VERSION_PHASES
+} from '../../utils/constants';
+import { getVersionData } from './utils';
+
+export const useTestPlanVersionsByPhase = testPlanVersions => {
+    const testPlanVersionsByPhase = useMemo(() => {
+        const initialPhases = Object.keys(TEST_PLAN_VERSION_PHASES).reduce(
+            (acc, key) => {
+                acc[TEST_PLAN_VERSION_PHASES[key]] = [];
+                return acc;
+            },
+            {}
+        );
+
+        return testPlanVersions.reduce((acc, testPlanVersion) => {
+            acc[testPlanVersion.phase].push(testPlanVersion);
+            return acc;
+        }, initialPhases);
+    }, [testPlanVersions]);
+
+    return { testPlanVersionsByPhase };
+};
+
+export const useDerivedOverallPhaseByTestPlanId = (
+    testPlans,
+    testPlanVersions
+) => {
+    const { testPlanVersionsByPhase } =
+        useTestPlanVersionsByPhase(testPlanVersions);
+
+    const getVersionDataByDirectory = (directory, phaseKey, phaseTimeKey) => {
+        if (
+            testPlanVersionsByPhase[phaseKey].some(
+                testPlanVersion =>
+                    testPlanVersion.testPlan.directory === directory
+            )
+        ) {
+            const { earliestVersion, latestVersion } = getVersionData(
+                testPlanVersionsByPhase[phaseKey],
+                phaseTimeKey
+            );
+            return phaseTimeKey ? earliestVersion?.phase : latestVersion?.phase;
+        }
+        return undefined;
+    };
+
+    const derivedOverallPhaseByTestPlanId = useMemo(() => {
+        const derivedOverallPhaseByTestPlanId = {};
+        const phases = [
+            {
+                phase: TEST_PLAN_VERSION_PHASES.RECOMMENDED,
+                timeKey: 'recommendedPhaseReachedAt'
+            },
+            {
+                phase: TEST_PLAN_VERSION_PHASES.CANDIDATE,
+                timeKey: 'candidatePhaseReachedAt'
+            },
+            {
+                phase: TEST_PLAN_VERSION_PHASES.DRAFT,
+                timeKey: 'draftPhaseReachedAt'
+            },
+            { phase: TEST_PLAN_VERSION_PHASES.RD }
+        ];
+        for (const testPlan of testPlans) {
+            for (const { phase, timeKey } of phases) {
+                const derivedPhase = getVersionDataByDirectory(
+                    testPlan.directory,
+                    phase,
+                    timeKey
+                );
+                if (derivedPhase) {
+                    derivedOverallPhaseByTestPlanId[testPlan.id] = derivedPhase;
+                    break;
+                }
+            }
+        }
+        return derivedOverallPhaseByTestPlanId;
+    }, [testPlans, testPlanVersions]);
+
+    return { derivedOverallPhaseByTestPlanId };
+};
+
+export const useTestPlansByPhase = (testPlans, testPlanVersions) => {
```

Comment:
I'm not 100% sure about this, and I think you should do whatever way you prefer, but here's a rough suggestion, which is to use plain functions instead of hook functions for these three internal helpers. It would interfere with the use of useMemo though. It's very subjective but the reason for that suggestion is that hooks require the reader to mentally track the reactivity of the hooks so it feels more intimidating to read.

Reply:
I can definitely see your point. My (not super strongly held) opinion is that since they are doing operations on potentially large arrays it is worth the trade off in readability to be able to avoid redoing those operations on unrelated re-renders. I am also hoping to one day get to refactor some code in DataManagement to use these internal helpers. They were designed to replace similar non-memo'd code in that component and DataManagementRow at a later date. I am just leaving that out of this PR for the sake of this feature making it before release cut-off.

But point definitely taken on not being inviting to readers. I wonder if this would be alleviated somewhat by each being in a dedicated file which is pretty standard practice. I am going to leave as-is for now but keep thinking on this.